src/c/keeper.c

```diff
@@ -0,0 +1,796 @@
+#include "keeper.h"
+#include "service.h"
+#include "errorlist.h"
+#include "bus.h"
+#include "devsdk/devsdk.h"
+#include "iot/time.h"
+#include "api.h"
+#include "iot/base64.h"
+
+
+/* Our "impl" structure for global state. */
+typedef struct keeper_impl_t 
+{
+    devsdk_service_t *service;
+    iot_threadpool_t *pool;
+    iot_logger_t *lc;
+    char *host;
+    uint16_t port;
+    char *key_root;
+    char *topic_root;
+    devsdk_registry_updatefn updater;
+    void *updatectx;
+} keeper_impl_t;
+
+static int32_t edgex_keeper_client_notify(void *impl, const iot_data_t *request, const iot_data_t *pathparams, const iot_data_t *params, iot_data_t **reply);
+#define CONF_PREFIX "edgex/v3/"
+
+void *devsdk_registry_keeper_alloc(devsdk_service_t *service)
+{
+    keeper_impl_t *rv;
+    rv = calloc(sizeof(keeper_impl_t), 1);
+    rv->service = service;
+    return rv;
+}
+
+static void edgex_keeper_client_free (void *impl)
+{
+  keeper_impl_t *keeper = (keeper_impl_t *)impl;
+  if (keeper)
+  {
+    if (keeper->host)
+    {
+        free (keeper->host);
+    }
+    if (keeper->key_root)
+    {
+        free(keeper->key_root);
+    }
+    if (keeper->topic_root)
+    {
+        free(keeper->topic_root);
+    }
+    if (impl)
+    {
+        free (impl);
+    }
+  }
+}
+
+static void *delayed_message_bus_connect(void *impl)
+{
+    keeper_impl_t *keeper = (keeper_impl_t *)impl;
+
+    iot_log_debug(keeper->lc, "Message bus wait thread starting");
+    while (true)
+    {
+        if (keeper->service && keeper->service->msgbus)
+        {
+            char *tree = malloc(strlen(keeper->topic_root) + 3);
+            strcpy(tree, keeper->topic_root);
+            if (tree[strlen(tree) - 1] == '/')
+            {
+                tree[strlen(tree) - 1] = '\0';
+            }
+            strcat(tree, "/#");
+            iot_log_info(keeper->lc, "Subscribing to Keeper notifications on message bus at %s", tree);
+            edgex_bus_register_handler(keeper->service->msgbus, tree, impl, edgex_keeper_client_notify);
```

Comment:
This can only subscribe to `edgex/configs/edgex/v4/{service_name}/#` topic. It also requires subscribing to `edgex/configs/edgex/v4/core-common-config-bootstrapper/all-services/#` and `edgex/configs/edgex/v4/core-common-config-bootstrapper/device-services/#`

Reply:
This will be addressed in #529.